system-helper/flatpak-system-helper.c

```diff
@@ -237,10 +339,53 @@ handle_deploy (FlatpakSystemHelper   *object,
   gboolean local_pull;
   gboolean reinstall;
   g_autofree char *url = NULL;
+  g_autoptr(OngoingPull) ongoing_pull = NULL;
+  g_autofree gchar *repo_path = NULL;
+  gint dir_fd = -1;
+
+  ongoing_pull = take_ongoing_pull_by_dir (arg_repo_path);
+  if (ongoing_pull != NULL)
+    {
+      g_autofree gchar *repo_basename = NULL;
+      g_autoptr(GError) local_error = NULL;
+
+      if (umount (ongoing_pull->mountpoint) == -1)
+        {
+          /* TODO: What else should do we here, other than error-reporting */
+          glnx_throw_errno_prefix (&local_error, "Failed to unmount dir %s", ongoing_pull->mountpoint);
+          g_warning ("%s", local_error->message);
+          g_clear_error (&local_error);
+        }
 
-  g_debug ("Deploy %s %u %s %s %s", arg_repo_path, arg_flags, arg_ref, arg_origin, arg_installation);
+      /* Killing the fuse backend will revoke all access to write operations. */
+      g_subprocess_force_exit (ongoing_pull->revokefs_backend);
+      g_clear_object (&ongoing_pull->revokefs_backend);
 
-  system = dir_get_system (arg_installation, get_sender_pid (invocation), &error);
+      repo_basename = g_path_get_basename (arg_repo_path);
+      repo_path = g_build_filename (ongoing_pull->src_dir, repo_basename, NULL);
+      repo_file = g_file_new_for_path (repo_path);
+
+      dir_fd = open (ongoing_pull->src_dir, O_DIRECTORY);
+      if (!flatpak_canonicalize_permissions (dir_fd,
```

Comment:
I think the revokefs writer process needs to explicitly strip setuid/setgid bits as well. Otherwise, user `wjt` can write an executable file with the setuid bit, wait for the revokefs writer to happily create it as user `flatpak`, then run it. This provides a way to write to files owned by `flatpak` user despite the revokefs write process being killed.

Reply:
Regarding this, I have been digging up details of what could be done to avoid this. The best approach I think is to teach revokefs'  `open()` callback to fail the `open()` operation if the request is coming in with `S_ISUID`/`S_ISGID` flags.

The problem for me while thinking this is that, there is no real notion of the file in revokefs. Everything is a fd, and read/write is occuring across fds with data buffers(and corresponding offset). Hence, to strip down setuid/setgid bits by conventional means (like fchmodat) make me fuzzy thinking about it (in the revokefs land). Thoughts/comments ?

Reply:
I don't understand your concern @uajain. We want to protect against creating files with setuid bits, so we'll have to check for them in all places where a mode argument is taken when modifying or creating files. This seems to me to be in writer.c: handle_mkdir(), handle_chmod() and handle_open(). Adding a bitmask check in these should be easy, no?